test/clustertest/tests/ClusterUpgradeTest.cpp

```diff
@@ -0,0 +1,172 @@
+
+        // Parse the JSON we got from github with recent releases.
+        string data = SFileLoad(tempJson);
+        SFileDelete(tempJson);
+        string bedrockTagName;
+        list<string> j1 = SParseJSONArray(STrim(data));
+        ASSERT_EQUAL(j1.size(), RECENT_RELEASES_TO_CHECK);
+
+        // We need the tag names from each release.
+        array<string, RECENT_RELEASES_TO_CHECK> tagNames;
+        for (size_t i = 0; i < RECENT_RELEASES_TO_CHECK; i++) {
+            STable j2 = SParseJSONObject(j1.front());
+            auto tag = j2.find("tag_name");
+            if (tag != j2.end()) {
+                tagNames[i] = tag->second;
+            }
+            if (j1.size()) {
+                j1.pop_front();
+            }
+        }
+
+        // Now choose the one to use. We want to test against the msot recent release that isn't the commit we're currently on.
+        // the commit number of the tag: git rev-list -n 1 $TAG
+        // The commit number we're currently on: git rev-parse HEAD
+        // If the current commit matches the tested tag, the script returns 1 and we check the next one. When the script returns 0, that's the release we'll use.
+        for (const auto& tagName : tagNames) {
+            string checkIfOnLatestTag = "/bin/bash -c 'if [ \"`git rev-list -n 1 " + tagName + "`\" = \"`git rev-parse HEAD`\" ]; then exit 1; else exit 0; fi'";
+            int result = system(checkIfOnLatestTag.c_str());
+            if (!result) {
+                bedrockTagName = tagName;
+                break;
+            }
+        }
+
+        // Make sure we got something to test.
+        ASSERT_NOT_EQUAL(bedrockTagName, "");
+
+        // If you'd like to test against a particular tag, uncomment the following line. The value chosen here was a
+        // known bad version that failed to escalate commands at upgrade when first deployed.
+        // bedrockTagName = "2022-05-06";
+
+        // If we've already built this, don't bother doing it again. This makes running this test multiple times in a
+        // row much faster.
+        prodBedrockName = "/tmp/bedrock-" + bedrockTagName; 
+        if (!SFileExists(prodBedrockName)) {
+            // Get a directory we can work in.
+            char brReleaseDirArr[] = "/tmp/br-prod-test-XXXXXX";
+            ASSERT_EQUAL(mkdtemp(brReleaseDirArr), brReleaseDirArr);
+            string brReleaseDir(brReleaseDirArr, sizeof(brReleaseDirArr) - 1);
+
+            // Clone bedrock.
+            ASSERT_FALSE(system(("cd " + brReleaseDir + " && git clone https://github.com/Expensify/Bedrock.git > /dev/null").c_str()));
+
+            // Check out the release tag.
+            ASSERT_FALSE(system(("cd " + brReleaseDir + " && cd Bedrock && git checkout " + bedrockTagName + "  > /dev/null").c_str()));
+
+            // Build the release.
+            ASSERT_FALSE(system(("cd " + brReleaseDir + " && cd Bedrock && make -j8 bedrock  > /dev/null").c_str()));
```

Comment:
bump

> > catch a very specific problem
> 
> This is what we normally do when adding tests, right?
> 
> > that we have only experienced once
> 
> This is both the median and mode number of times we experience a problem that breaks production before adding tests to catch it happening again, right?
> 
> > I think it's unlikely to catch issues in the future
> 
> I disagree, I think this is highly likely to catch issues in the future and it's why I want to add it now, before removing `ESCALATE` and the related commands, because it will catch me breaking upgrades by doing this.
> 
> > Would this have caught the other http escalation issues from this migration?
> 
> I'm not 100% sure without looking back over those, but I think likely yes, for at least some of them. We have for sure broken upgrades in the past, and have also carefully orchestrated "this needs to go out first before this can go out which removes a thing that the first change makes obsolete" many times. This would catch any time we should have done that dance but missed it.

ok, I agree this should cover at least some reasonable class of errors we would want to prevent. Bump on my review question about using the release artifacts instead of local compilation.